infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,188 @@
+                    // from the trim mark instead of the trailing committed tail.
+                    if (committedTail < trimMark.getSequence() - 1) {
+                        Token trimToken = new Token(trimMark.getEpoch(), trimMark.getSequence() - 1);
+                        getCorfuRuntime().getAddressSpaceView().prefixTrim(trimToken, false);
+                        committedTail = trimMark.getSequence() - 1;
+                    }
+                    lastEpoch = currEpoch;
+                }
+
+                log.debug("runAutoCommit: trying to commit [{}, {}].", committedTail + 1, lastLogTail);
+
+                // Commit addresses in batches, retry limit is shared by all batches in this cycle.
+                // NOTE: This implementation relies on the fact that state transfer invokes the
+                // read protocol and fill holes, otherwise the committedTail could be invalid.
+                // (e.g. 1. State transferred a hole at address 100 from A to B; 2. Commit address
+                // 100, which only goes to A as B is not in this address segment; 3. B finishes
+                // state transfer and segments merged; 4. Send a new committedTail to B which is
+                // invalid as 100 is still a hole on B.
+                while (committedTail < lastLogTail) {
+                    long commitStart = committedTail + 1;
+                    long commitEnd = Math.min(committedTail + COMMIT_BATCH_SIZE, lastLogTail);
+                    getCorfuRuntime().getAddressSpaceView().commit(commitStart, commitEnd);
+                    log.trace("runAutoCommit: successfully committed batch [{}, {}]", committedTail, commitEnd);
+                    committedTail = commitEnd;
+                }
+
+                log.debug("runAutoCommit: successfully finished auto commit cycle. " +
+                        "New committed tail: {}.", committedTail);
+                break;
+
+            } catch (RuntimeException re) {
+                log.warn("runAutoCommit: encountered an exception on attempt {}/{}.",
+                        i, MAX_COMMIT_RETRY, re);
+
+                if (i >= MAX_COMMIT_RETRY) {
+                    log.warn("runAutoCommit: retry exhausted, abort and wait for next cycle");
+                }
+
+                Throwable cause = re.getCause();
+                if (cause instanceof TimeoutException || cause instanceof NetworkException) {
+                    Sleep.sleepUninterruptibly(CONN_RETRY_RATE);
+                }
+
+            } catch (Throwable t) {
+                log.error("runAutoCommit: encountered unexpected exception", t);
+                lastLogTail = getCorfuRuntime().getSequencerView().query().getSequence();
+                throw t;
+            }
+        }
+
+        // Update lastLogTail no matter commit cycle succeeds or not.
+        lastLogTail = getCorfuRuntime().getSequencerView().query().getSequence();
```

Comment:
Right now when we get a token from a sequencer we do it via layoutHelper, which retries up to systemDownHandlerTriggerLimit if something goes wrong. In some cases, for example, when the sequencer requires a full bootstrap from a failure detector, this can get stuck for quite a long period of time. I am wondering if there is a possibility of this interfering with a periodically scheduled auto-commit task. For example, if this update to the lastLogTail can somehow interfere with a next scheduled runAutoCommit task because it takes longer than expected to finish it.

Reply:
I changed to use seqeuncer client instead of sequencer view